src/main/java/org/jenkinsci/plugins/workflow/cps/CpsThread.java

```diff
@@ -135,6 +135,16 @@ public ContextVariableSet getContextVariables() {
         return contextVariables;
     }
 
+    /**
+     * Used to remove context variables after a {@link CpsBodyExecution} finishes executing so
+     * that those variables will no longer be persisted as part of the program. See JENKINS-53709.
+     */
+    void popContextVariables() {
+        if (contextVariables != null) {
+            contextVariables = contextVariables.getParent();
```

Comment:
I think it makes sense to treat the variables like a stack and pop them to get the parent variables rather than just nulling them out, but I am not entirely sure.

Reply:
I think there is just a different `CpsThread` in that case. If this were not true, there would already be a more visible bug (not just a leak). Possibly suffices to just

```java
contextVariables = null;
```

You could check by having some block-scoped step bind a context variable, then having another step run after its body completes check whether that variable is still accessible.

Reply:
I did try something like the following before my change, and it failed because the `sh` step cannot find the context variables as expected. Is that what you mean?

    node {
        // Binds FilePath, Executor, TaskListener, context variables (I think)
    }
    sh('echo "foo"') // Context variables should be gone here